### DISCUSSION TITLE
Tab Rules proposal

### DISCUSSION THREAD

**Discussion by ssorallen** [tabwrangler/tabwrangler#358] [General]
Consolidate the various conditions for closing and preserving tabs into a single concept called "Actions". A rule in concept is the following:

* If a tab meets condition X do action Y

Current Tab Wrangler functionality can all fit into Rules (formerly Actions). Currently these conditions are checked from top to bottom, once every 5 seconds. That 5 second "tick" would potentially need to remain. The following list is current features in the "Options" tab, checked from top to bottom on every tick:

1. If the tab has been active for 1s, reset its counter
2. If the tab is playing audio, lock it
3. If the tab is in a group, lock it
4. If the tab's URL matches a custom string, lock it
5. If the tab's counter is longer than period X, close it

This framework would enable request features like the following:

* Auto close a tab if title matches regex: https://github.com/tabwrangler/tabwrangler/issues/351

**Comment by ssorallen** [comment]
@KTamas This is the concept I had been considering. The UI/UX for this would be a consolidated table on the "Options" tab that shows each Action.

**Reply by ssorallen** [reply]
This needs more thought before trying to work on it. Let me know what you think. The code for closing tabs already sort of works this way, and moving more toward a single "Action" concept would enable a greater number of custom features.

**Reply by sarahkadar** [reply]
@ssorallen It's intriguing, though first I'd like to triage all the tickets and see where we're at.

**Comment by jacobSingh** [comment]
This is an interesting idea and I could see some power users really
appreciating a full workflow interface like this.  Since I don't do shit
anymore, I don't really get a vote, but IMHO, it should be in an "advanced"
tab or something since the majority of users would struggle to figure out
how to use it.

So the basic tab just operates as it would normally, and it would create a
couple simple "actions" (I prefer the term "rules" personally), but that
machinery would be hidden.  Anyone who goes to "advanced" and modifies the
core actions would no longer be able to use the basic view any more.

Best,
J

On Fri, Dec 3, 2021 at 8:17 PM Ross Allen ***@***.***> wrote:

> This needs more thought before trying to work on it. Let me know what you
> think. The code for closing tabs already sort of works this way, and moving
> more toward a single "Action" concept would enable a greater number of
> custom features.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tabwrangler/tabwrangler/discussions/358#discussioncomment-1744409>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA5ZBWREECBUGAXORZHO3DUPDKA3ANCNFSM5JJ26MKQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
+919971876580
twitter: @JacobSingh ( http://twitter.com/#!/JacobSingh )
web: http://www.jacobsingh.name
Skype: pajamadesign
gTalk: ***@***.***

**Comment by ssorallen** [comment]
Issues I am considering as part of this implementation the following issues:
* https://github.com/tabwrangler/tabwrangler/issues/435 (rule priorities)
* https://github.com/tabwrangler/tabwrangler/issues/242 (time periods when rules do and do not apply)
* https://github.com/tabwrangler/tabwrangler/issues/490 (rule sets, toggle between different rule sets)
* https://github.com/tabwrangler/tabwrangler/issues/378 (rule ordering, put "close" rules at the top to apply first)
* https://github.com/tabwrangler/tabwrangler/issues/255 (rules can have different durations)

This is note taking at the moment, and I will come up with a planning document eventually.

**Comment by ssorallen** [comment]
I added my initial write up (that I made in collaboration with Claude Code) as a wiki page so I can continue to edit it: https://github.com/tabwrangler/tabwrangler/wiki/Tab-Rules-RFC
